python/ray/includes/global_state_accessor.pxd

```diff
@@ -46,6 +46,65 @@ cdef extern from "ray/gcs/gcs_client/global_state_accessor.h" nogil:
             const c_string &node_ip_address,
             c_string *node_to_connect)
 
+cdef extern from * namespace "ray::gcs" nogil:
+    """
+    #include <thread>
+    #include "ray/gcs/gcs_server/store_client_kv.h"
+    namespace ray {
+    namespace gcs {
+
+    bool RedisGetKeySync(const std::string& host,
+                         int32_t port,
+                         const std::string& password,
+                         bool use_ssl,
+                         const std::string& config,
+                         const std::string& key,
+                         std::string* data) {
+      RedisClientOptions options(host, port, password, false, use_ssl);
+
+      std::string config_list;
+      RAY_CHECK(absl::Base64Unescape(config, &config_list));
+      RayConfig::instance().initialize(config_list);
+
+      instrumented_io_context io_service;
+
+      auto redis_client = std::make_shared<RedisClient>(options);
+      auto status = redis_client->Connect(io_service);
+
+      if(!status.ok()) {
+        RAY_LOG(ERROR) << "Failed to connect to redis: " << status.ToString();
+        return false;
+      }
+      auto cli = std::make_unique<StoreClientInternalKV>(
+        std::make_unique<RedisStoreClient>(std::move(redis_client)));
+
+      bool ret_val = false;
+      cli->Get("session", key, [&](std::optional<std::string> result) {
+        if (result.has_value()) {
+          *data = result.value();
+          ret_val = true;
+        } else {
+          RAY_LOG(ERROR) << "Failed to get " << key;
+          ret_val = false;
+        }
+      });
+      io_service.run_for(std::chrono::milliseconds(1000));
```

Comment:
This means it is 1 second timeout?

Reply:
Yep :)

Reply:
Actually, I am a little confused about the behavior. If this returns without invoking the callback, is callback going to be canceled when it actually comes?

Reply:
So basically once the io_context stops running, the remaining callbacks will only continue the next time the io_context runs (indefinitely queued). But this one is a local variable, so it will just go out of scope. Even if we use the same thread for different io_contexts later, I don't think they will actually execute the callback unless they queue it again.

Reply:
I feel if the redis failed, the driver will crash (https://github.com/ray-project/ray/blob/master/src/ray/gcs/redis_context.cc#L176). This is bad, but fine for now.

I feel here 1s here might not be good. It's too small. I'm not sure whether the crash is executed in redis threads or not.

Let's maybe set it up according to https://github.com/ray-project/ray/blob/master/src/ray/common/ray_config_def.h#L324-L331

Reply:
Btw, if we set run_for for 5 seconds, and if the callback is called in 0.1 second, does it finish in 0.1 second or still wait. 5 seconds?

Reply:
1. I agree with iycheng. 1s might be too short when redis‘s CPU busy.
2.  why we don't use the Python Redis client directly ？